pkg/server/server.go

```diff
@@ -0,0 +1,182 @@
+)
+
+type Server struct {
+	redisClient RedisClient
+	router      Router
+}
+
+type Request struct {
+	RetQueue string `json:"ret_queue"`
+	Now      uint64 `json:"now"`
+	Cmd      string `json:"cmd"`
+	Data     string `json:"data"`
+}
+
+type Response struct {
+	Result string `json:"result"`
+	Err    error  `json:"error"`
+}
+
+type RedisClient struct {
+	Pool *redis.Pool
+}
+
+func NewServer() (*Server, error) {
+	redisClient, err := NewRedisClient()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create new redis client")
+	}
+
+	r := Router{
+		routes: make(map[string]func(ctx context.Context, data string) (string, error)),
+	}
+
+	server := Server{
+		redisClient,
+		r,
+	}
+
+	server.Register("login", router.Login)
+
+	server.Register("machines.deploy", router.MachinesDeploy)
+	server.Register("machines.get", router.MachinesGet)
+	server.Register("machines.delete", router.MachinesDelete)
+	server.Register("machines.machine.add", router.MachineAdd)
+	server.Register("machines.machine.remove", router.MachineRemove)
+
+	server.Register("gateway.name.deploy", router.GatewayNameDeploy)
+	server.Register("gateawy.name.get", router.GatewayNameGet)
+	server.Register("gateway.name.delete", router.GatewayNameDelete)
+	server.Register("gateway.fqdn.deploy", router.GatewayFQDNDeploy)
+	server.Register("gateway.fqdn.get", router.GatewayFQDNGet)
+	server.Register("gateway.fqdn.delete", router.GatewayFQDNDelete)
+
+	server.Register("k8s.get", router.K8sGet)
+	server.Register("k8s.deploy", router.K8sDeploy)
+	server.Register("k8s.delete", router.K8sDelete)
+	server.Register("k8s.node.add", router.K8sAddNode)
+	server.Register("k8s.node.delete", router.K8sRemoveNode)
+
+	server.Register("zdb.deploy", router.ZDBDeploy)
+	server.Register("zdb.delete", router.ZDBDelete)
+	server.Register("zdb.get", router.ZDBGet)
+
+	return &server, nil
+}
+
+func (s *Server) Register(route string, fn func(context.Context, string) (string, error)) {
+	s.router.routes[route] = fn
+}
+
+func NewRedisClient() (RedisClient, error) {
+	pool, err := newRedisPool()
+	if err != nil {
+		return RedisClient{}, errors.Wrap(err, "failed to create new redis pool")
+	}
+	return RedisClient{
+		Pool: pool,
+	}, nil
+}
+
+func newRedisPool() (*redis.Pool, error) {
+	return &redis.Pool{
+		Dial: func() (redis.Conn, error) {
+			return redis.Dial("tcp", "localhost:6379")
+		},
+		TestOnBorrow: func(c redis.Conn, t time.Time) error {
+			if time.Since(t) > 10*time.Second {
+				_, err := c.Do("PING")
+				return err
+			}
+
+			return nil
+		},
+		MaxActive:   100,
+		IdleTimeout: 1 * time.Minute,
+		Wait:        true,
+	}, nil
+}
+
+// Run watches a redis queue for incoming messages
+func (s *Server) Run(ctx context.Context) error {
+	con := s.redisClient.Pool.Get()
+	defer con.Close()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			res, err := redis.ByteSlices(con.Do("BRPOP", "tfgrid.client", 0))
```

Comment:
This will wait indefinitely for a pop. Could be good to allow closing the server... You could add a variable running to Server and whenever that is false you break out of this loop. Then the timeout value here could be some value so that you break out of the loop once in a while to be able to shutdown properly. It will probably help for testing purposes too.

Reply:
i agree with you on the timeout, it certainly should be added.
for the closing the server part, i think there are two options:
1- just kill the process :D
2- pass a context that can be canceled to server.Run(), this way whenever the context is canceled, run should exit and the server stops

Reply:
You can do this for example: 
```
for s.running {
        select {
		case <-ctx.Done():
		        // I think this can be used to cancel running
			return nil
		default:
		        // provide a timeout of 1 second on brpop, the timeout I think will throw an error but just ignore that error, we will then go into the next loop which should either get back here or return the function if we closed the context
			res, err := redis.ByteSlices(con.Do("BRPOP", "tfgrid.client", 1))
      .....
			
}
```